### PR TITLE
Import code from the project basisLieHighestWeight

### PR DESCRIPTION
**Pull request by BenWilop**

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ version = "0.12.0-DEV"
 [deps]
 AbstractAlgebra = "c3fe647b-3220-5bb0-a1ea-a7954cac585d"
 AlgebraicSolving = "66b61cbe-0446-4d5d-9090-1ff510639f9d"
+Distributed = "8ba89e20-285c-5b6f-9357-94700520ee1b"
```
Comment (fingolfin) [inline Project.toml]: Why?
```suggestion
```

File: experimental/basisLieHighestWeight/VectorSpaceBases.jl [hunk 2]
```diff
@@ -0,0 +1,93 @@
+# manages the basisvectors and its linear independence
+
+using Oscar
+using SparseArrays
+
+ZZ = Int
+TVec = SparseVector{ZZ, Int} #--- Werte sind ZZ, Indizies sind Int (TVec ist Datentyp der Basisvektoren in basisLieHighestWeight)
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/VectorSpaceBases.jl]: German comments are no-go

File: experimental/basisLieHighestWeight/TensorModels.jl [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        end
+        wi = Int.(1:length(hw) .== i) # i-th fundamental weight
+        _mats = matricesForOperators(L, wi, ops)
+        _mats = tensorPowers(_mats, hw[i])
+        mats = mats == [] ? _mats : tensorProducts(mats, _mats)
+        #display(mats)
+    end
+
+    return mats
+end
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/TensorModels.jl]: Please make sure your source files end with a newline, e.g. by configured your editor to honor our `.editorconfig`, as described in the OSCAR Developer Documentation.

File: experimental/basisLieHighestWeight/MB.jl [hunk 4]
```diff
@@ -0,0 +1,364 @@
+module MB
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/MB.jl]: Why `MB`?

File: experimental/basisLieHighestWeight/LieAlgebras.jl [hunk 5]
```diff
@@ -0,0 +1,50 @@
+# ?
+
+using Oscar
+using SparseArrays
+
+G = Oscar.GAP.Globals
+forGap = Oscar.GAP.julia_to_gap
+fromGap = Oscar.GAP.gap_to_julia
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/LieAlgebras.jl]: Don't do that:
- you have this exact code in multiple files, that's bad
- it makes it very hard for others to read your code, as locally it is not clear what `G` stands for
- `gap_to_julia` shouldn't be used in production code, it is just for interactive use, as it is not type stable. Instead specify explicit target types, e.g. `Vector{Int}(gapobj)` to convert a GAP list of integers into a `Vector{Int}`, etc. (this is type stable and at the same time also verifies the GAP object has the right shape, and if not, throws an exception)
- don't use `julia_to_gap` either, instead use `GAP.Obj`

File: Project.toml [hunk 6]
```diff
@@ -19,6 +20,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 RandomExtensions = "fb686558-2515-59ef-acaa-46db3789a887"
 RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
 Singular = "bcd08a7b-43d2-5ff7-b6d4-c458787f915c"
+SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
```
Comment (fingolfin) [inline Project.toml]: I am somewhat reluctant to add a depdency on `SparseArrays` as we have our own sparse matrix type. But I guess it depends on what you use it for, which is not clear to me.

File: experimental/basisLieHighestWeight/LieAlgebras.jl [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+
+function lieAlgebra(t::String, n::Int)
+    """
+    Creates the Lie-algebra as a GAP object that gets used for a lot other computations with GAP
+    """
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/LieAlgebras.jl]: Docstrings must come before the function, not inside it.

File: experimental/basisLieHighestWeight/MB.jl [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+export basisLieHighestWeight2
+export is_fundamental
+
+using Polymake
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/MB.jl]: What is this needed for?

File: experimental/basisLieHighestWeight/MB.jl [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+        vec = calc_vec(v0, mon, mats)
+        wt = calc_wt(mon, wts)  
+        #println("vec:" , vec)
+        #println("wt: ", wt)
+        if !haskey(space, wt)
+            space[wt] = nullSpace()
+        end
+        addAndReduce!(space[wt], vec)
+    end
+    GC.gc()
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/MB.jl]: Why?

File: experimental/basisLieHighestWeight/MonomialOrder.jl [hunk 10]
```diff
@@ -0,0 +1,110 @@
+# manages methods for different orders of monomials
```
Comment (fingolfin) [inline experimental/basisLieHighestWeight/MonomialOrder.jl]: OSCAR already has a system of monomial orderings, any reason you are not using that?

**Review by fingolfin — CHANGES_REQUESTED** [review]
Hi there, thanks for your contribution! Could  you elaborate a bit about the background of this, and also your own? I.e, what are you trying to achieve here? Are you working for someone involved in the OSCAR project?

Unfortunately this code has substantial issues. In its current form, it is unsuitable for merging, not even into `experimental`. But if you are interested in improving it: I left a bunch of comments indicating major issues. Once those are addressed, I'd be happy to go again over the code and point out less severe issues.

**Comment by lkastner** [comment]
> Hi there, thanks for your contribution! Could you elaborate a bit about the background of this, and also your own? I.e, what are you trying to achieve here? Are you working for someone involved in the OSCAR project?
> 
We opened this PR so we can work on this together and have the tests run. It will take a while to straighten out, hence it is a draft. We will also add a better description later.

**Comment by thofma** [comment]
@lkastner In view of the PBWDeformations.jl things, do you think it would make sense to call these `lieAlgebra` functions here `gap_lie_algebra` or some similar abomination? Then we could cleanly import the stuff from PBWDeformations.jl. But maybe you had a different plan for the `lieAlgebra(...)` functions? I was just wondering how to do this without too much refactoring.

**Comment by BenWilop** [comment]
Hi everyone, thanks for all the comments! I am working as a student assistant for Ghislain Fourier in Aachen. Lars and Mara are not responsible for the project and only helped me getting started with the integration. I took over the project 1 year ago and did not change design choices that probably happened with regards to an older status of Oscar. I will try to fix all issues that come up. There are many of parts of the code that I did not clean up (such as comments) and I will do that as well.

File: experimental/basisLieHighestWeight/MB.jl (hunk 4 above)
Comment (BenWilop) [reply experimental/basisLieHighestWeight/MB.jl]: MB stand for MonomialBasis, but I changed it to BasisLieHighestWeight now.

File: Project.toml (hunk 6 above)
Comment (BenWilop) [reply Project.toml]: This changed now and the code uses only the Oscar-Type now. But there are operations that are not supported in Oscar, such as SparseMatrix*SparseMatrix and the Kronecker-Product.

File: experimental/basisLieHighestWeight/MB.jl (hunk 8 above)
Comment (BenWilop) [reply experimental/basisLieHighestWeight/MB.jl]: The user is only supposed to use the two exported functions, the rest is part of the algorithm.

File: experimental/basisLieHighestWeight/MB.jl (hunk 9 above)
Comment (BenWilop) [reply experimental/basisLieHighestWeight/MB.jl]: I don't know why, but the garbage collector of Julia sometimes doesn't delete the dereferenced objects of the function add_known_monomial and add_new_monomials. Therefore, I had memory problems without calling it manually. I did not yet find a minimal working example and can't figure out the cause. After completing the other major changes, I will test it again and try to find another solution.

File: experimental/basisLieHighestWeight/MonomialOrder.jl (hunk 10 above)
Comment (BenWilop) [reply experimental/basisLieHighestWeight/MonomialOrder.jl]: I will switch to the Julia-Type, but both an older version of the code and parts of the code that create monomials through integer points of a polytope made it easier for me to keep the integers.

File: Project.toml (hunk 6 above)
Comment (fingolfin) [reply Project.toml]: I am confused; what does "This changed now" mean? `SparseArrays` still is listed here (as is `Distributed`)

File: test/Experimental/basisLieHighestWeight/MBOld.jl [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+
+    return v
+end
+
+
+#### Lie algebras
+
+G = Oscar.GAP.Globals
+forGap = Oscar.GAP.julia_to_gap
+fromGap = Oscar.GAP.gap_to_julia
```
Comment (fingolfin) [inline test/Experimental/basisLieHighestWeight/MBOld.jl]: Just to say, there are still multiple copies of the above three "aliases", and as I pointed out before, they should all go, as should all uses of `Oscar.GAP.julia_to_gap
` and `Oscar.GAP.gap_to_julia`. But I am guessing you have this on your TODO and just did not yet get to it.

File: test/Experimental/basisLieHighestWeight/MBOld.jl [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
+    return length(res[1]), res...
+end
+
+nullMon(m) = zeros(Short, m)
+
+
+function compute(v0, mats, wts::Vector{Vector{Int}})
+    m = length(mats) #---
+    monomials = [nullMon(m)]    #--- hier speichern wir die Monombasis. Ein Eintrag steht für Potenzen unserer Basiselemente
```
Comment (fingolfin) [inline test/Experimental/basisLieHighestWeight/MBOld.jl]: more Ger

**Comment by lkastner** [comment]
I wanted to rebase this branch onto master, but there is stuff hidden in some merge commits (msg "substituted Package SparseArray with Oscar SMat"). So as the only way out now I see the creation of a new branch and then overwrite this branch by force pushing to it. Would that work for you @BenWilop @marabelotti ?

**Comment by fingolfin** [comment]
@lkastner why is that the only way out? can't you just merge `master` into this branch? We will mostly likely squash-merge in the end anyway, so there is no need for a super clean history, is there?

**Comment by fingolfin** [comment]
(not that I mind either way *shrug*)

**Comment by lkastner** [comment]
> @lkastner why is that the only way out? can't you just merge `master` into this branch? We will mostly likely squash-merge in the end anyway, so there is no need for a super clean history, is there?

Ah, yes, I am so used to rebasing these days. I'll just do that.

**Comment by fingolfin** [comment]
Is there already a plan how to proceed with this PR, in light of the Lie algebra work by @lgoettgens that was already merged?

**Comment by lgoettgens** [comment]
Besides the initial work in #2145, there is already a plan in #2207 to bring a connection between Oscar and GAP Lie algebras. This would make it possible to, similar to GAPGrroups, handle Oscar Lie algebras in your code and delegate computations to GAP using isomorphisms that will be introduced in #2207.

**Comment by BenWilop** [comment]
Okay great. The only part where Lie-Algebras get used is in the file LieAlgebras.jl to call GAP-functions. I think I will wait for  #2207 and then use it 👍 . 

Furthermore, the problems of the PR mentioned so far are fixed and I wanted to ask for another review. I don't understand all of the mathematical background of the project and since answers to my questions were simplified / maybe didn't use the correct vocabulary, the doc of the algorithm basis_lie_highest_weight is still rather short. From my point of view the only relevant algorithm is basis_lie_highest_weight and everything else was only implemented for it and not with the intention to get integrated into Oscar. If there is use for other parts of the code, I can split it up. But maybe it will make more sense to wait with this until everything else Lie-Algebra related is finished.

**Comment by fingolfin** [comment]
PR #2207 now has been merged :-)

**Comment by lgoettgens** [comment]
@BenWilop an I will try to set up a meeting next week to continue working on this.

**Comment by lgoettgens** [comment]
@BenWilop and I are trying to set up a meeting for next week to continue working on this together.

**Comment by BenWilop** [comment]
Closed in favour of Pull request https://github.com/oscar-system/Oscar.jl/pull/2402. We opened a new one because of access rights.

**Pull request closed without merging**